plugins/logging/pkg/util/generic.go

```diff
@@ -0,0 +1,45 @@
+package util
+
+import "sync"
+
+type AsyncClient[T any] struct {
+	Client      T
+	initCond    *sync.Cond
+	initialized bool
+}
+
+func NewAsyncClient[T any]() *AsyncClient[T] {
+	return &AsyncClient[T]{
+		initCond: sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+func (c *AsyncClient[T]) WaitForInit() {
+	c.initCond.L.Lock()
+	for !c.initialized {
+		c.initCond.Wait()
+	}
+	c.initCond.L.Unlock()
+}
+
+func (c *AsyncClient[T]) IsInitialized() bool {
+	c.initCond.L.Lock()
+	defer c.initCond.L.Unlock()
+	return c.initialized
+}
+
+func (c *AsyncClient[T]) SetClient(setter func() T, alwaysSet bool) {
+	c.initCond.L.Lock()
+	defer c.initCond.L.Unlock()
+
+	if c.initialized {
+		if !alwaysSet {
+			return
+		}
+		c.Client = setter()
+		return
+	}
+	c.Client = setter()
+	c.initialized = true
+	c.initCond.Broadcast()
+}
```

Comment:
This logic doesn't really make sense in the context of AsyncClient as a whole. Perhaps this needs to be separate functions or separate types.